flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java

```diff
@@ -0,0 +1,186 @@
+			Preconditions.checkState(results.size() < maxCheckpoints, "results.size() > maxCheckpoints", results.size(), maxCheckpoints);
+			enqueue(new CheckpointStartRequest(checkpointId, result, checkpointOptions.getTargetLocation()), false);
+			return result;
+		});
+		Preconditions.checkArgument(put == result, "result future already present for checkpoint id: " + checkpointId);
+	}
+
+	@Override
+	public void addInputData(long checkpointId, InputChannelInfo info, int startSeqNum, Buffer... data) {
+		LOG.debug("add input data, checkpoint id: {}, channel: {}, startSeqNum: {}, num buffers: {}",
+			checkpointId, info, startSeqNum, data == null ? 0 : data.length);
+		enqueue(write(checkpointId, info, checkBufferType(data)), false);
+	}
+
+	@Override
+	public void addOutputData(long checkpointId, ResultSubpartitionInfo info, int startSeqNum, Buffer... data) {
+		LOG.debug("add output data, checkpoint id: {}, channel: {}, startSeqNum: {}, num buffers: {}",
+			checkpointId, info, startSeqNum, data == null ? 0 : data.length);
+		enqueue(write(checkpointId, info, checkBufferType(data)), false);
+	}
+
+	@Override
+	public void finishInput(long checkpointId) {
+		LOG.debug("finish input data, checkpoint id: {}", checkpointId);
+		enqueue(completeInput(checkpointId), false);
+	}
+
+	@Override
+	public void finishOutput(long checkpointId) {
+		LOG.debug("finish output data, checkpoint id: {}", checkpointId);
+		enqueue(completeOutput(checkpointId), false);
+	}
+
+	@Override
+	public void abort(long checkpointId, Throwable cause) {
+		LOG.debug("abort, checkpoint id: {}", checkpointId);
+		enqueue(ChannelStateWriteRequest.abort(checkpointId, cause), true); // abort already started
+		enqueue(ChannelStateWriteRequest.abort(checkpointId, cause), false); // abort enqueued but not started
+	}
+
+	@Override
+	public ChannelStateWriteResult getWriteResult(long checkpointId) {
+		LOG.debug("requested write result, checkpoint id: {}", checkpointId);
+		ChannelStateWriteResult result = results.remove(checkpointId);
+		Preconditions.checkArgument(result != null, "channel state write result not found for checkpoint id " + checkpointId);
+		return result;
+	}
+
+	public void open() {
+		executor.start();
+	}
+
+	@Override
+	public void close() throws IOException {
+		results.clear();
+		executor.close();
+	}
+
+	private void enqueue(ChannelStateWriteRequest request, boolean atTheFront) {
+		// state check and previous errors check are performed inside the worker
+		try {
+			if (atTheFront) {
+				executor.submitPriority(request);
+			} else {
+				executor.submit(request);
+			}
+		} catch (Exception e) {
+			request.cancel(e);
+			throw new RuntimeException("unable to send request to worker", e);
+		}
+	}
+
+	private static Buffer[] checkBufferType(Buffer... data) {
+		if (data == null) {
+			return new Buffer[0];
+		}
+		try {
+			for (Buffer buffer : data) {
+				Preconditions.checkArgument(buffer.isBuffer());
```

Comment:
Since we encountered a hard error in Flink, there is no way to proceed and I'd probably also switch `checkArgument` to `checkState`.

Reply:
The class itself (and its "downstream" classes) is still operable. It shouldn't know that it's clients aren't able to deal with this exception.